py_zipkin/request_helpers.py

```diff
@@ -0,0 +1,147 @@
+
+log = logging.getLogger(__name__)
+
+
+def _parse_single_header(b3_header):
+    """
+    Parse out and return the data necessary for generating ZipkinAttrs.
+
+    Returns a dict with the following keys:
+        'trace_id':             str or None
+        'span_id':              str or None
+        'parent_span_id':       str or None
+        'sampled_str':          '0', '1', 'd', or None
+        'sample_decision_only': True if header contains only a sampling
+                                decision, false-y value otherwise
+    """
+    parsed = dict.fromkeys(('trace_id', 'span_id', 'parent_span_id',
+                            'sampled_str', 'sample_decision_only'))
+
+    # b3={TraceId}-{SpanId}-{SamplingState}-{ParentSpanId}
+    #      (last 2 fields optional)
+    # OR
+    # b3={SamplingState}
+    bits = b3_header.split('-')
+
+    # Handle the lone-sampling-decision case:
+    if len(bits) == 1:
+        parsed['sampled_str'] = bits[0]
+        parsed['sample_decision_only'] = True
+        return parsed
+    if len(bits) > 4:
+        # Too many segments, just return the initialized defaults which will
+        # translate to a None result (no valid incoming span context or
+        # sampling decision)
+        return parsed
+    if len(bits) > 3:
+        parsed['parent_span_id'] = bits[3]
+    if len(bits) > 2:
+        parsed['sampled_str'] = bits[2]
+    if len(bits) > 1:
+        parsed['span_id'] = bits[1]
+        parsed['trace_id'] = bits[0]
+    return parsed
+
+
+def _parse_multi_header(headers):
+    """
+    Parse out and return the data necessary for generating ZipkinAttrs.
+
+    Returns a dict with the following keys:
+        'trace_id':             str or None
+        'span_id':              str or None
+        'parent_span_id':       str or None
+        'sampled_str':          '0', '1', 'd', or None
+        'sample_decision_only': True if header contains only a sampling
+                                decision, false-y value otherwise
+    """
+    parsed = {
+        'trace_id': headers.get('X-B3-TraceId', None),
+        'span_id': headers.get('X-B3-SpanId', None),
+        'parent_span_id': headers.get('X-B3-ParentSpanId', None),
+        'sampled_str': headers.get('X-B3-Sampled', None),
+        'sample_decision_only': False,
+    }
+    if headers.get('X-B3-Flags') == '1':
+        parsed['sampled_str'] = 'd'
+    # Unlike the single-header, "absent" and "present-but-empty-string" are
+    # both representable, and the spec only mentions Absent for encoding
+    # "Defer".  So that means we need to translate an empty-string to a
+    # non-empty, but invalid string.
+    if parsed['sampled_str'] == '':
+        parsed['sampled_str'] = 'INVALID'
+    parsed['sample_decision_only'] = (
+        all(parsed[k] is None for k in ('trace_id', 'span_id', 'parent_span_id'))
+        and parsed['sampled_str'])
+    return parsed
+
+
+def extract_zipkin_attrs_from_headers(headers, sample_rate=100.0,
```

Comment:
why does extracting use a sample rate? The result of extracting headers is just the data extracted.. turning the headers into a new context sure.. but why are the concerns conflated here?

Reply:
Well, this is my call-site, which I'm pretty happy with:
```
# Note: this is a monkey-patch for eventlet.wsgi.HttpProtocol.handle_one_response
# Also api.sample_rate_pct is a global for the process set via config file
def _patched_handle_one_response(self):
    zipkin_attrs = api.extract_zipkin_attrs_from_headers(
        self.headers, sample_rate=api.sample_rate_pct, use_128bit_trace_id=True)
    ...
    with api.ezipkin_server_span(
        service_name=api.default_service_name(),
        span_name=self.command,
        zipkin_attrs=zipkin_attrs,
        sample_rate=None if zipkin_attrs else api.sample_rate_pct,
...
    ) as zipkin_span:
        ...
```
Is there something undesirable there that I'm missing?

Reply:
in other codebases, we reuse code like you've written in scenarios that don't assume a specific sampling rate implementation. To make things more flexible, we compose the feature to extract the headers with whatever it is that needs it (ex applying a rate). By keeping these things separate, we can also clarify the unit tests which have little need to concern themselves with rate.

Reply:
Interesting.

Like you, I don't maintain this codebase either.  But this is a new helper-function.  In the past, anyone using `py_zipkin` had to write their own functions to do this.  So no existing users will be hampered by any lack of flexible composition here.

I think anyone needing that use-case could perform follow-on work here to add that complexity when it's necessary.

One of the things I liked about `py_zipkin` when looking for a python tracing library was how it did nearly all I wanted it to (subject to a couple open PRs, hehe), and not much more.